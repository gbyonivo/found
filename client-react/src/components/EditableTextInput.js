import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { BiX, BiPencil } from 'react-icons/bi';
import { TiTick } from 'react-icons/ti';

const className = 'bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold p-2 rounded inline-flex items-center ml-4';

function EditableTextInput({
  defaultValue, save, clear, label, type, fixed, valueKey,
}) {
  const [disabled, setDisabled] = React.useState(true);
  const [value, setValue] = React.useState(defaultValue);
  const onChange = ({ target }) => {
    setValue(target.value)
  };
  const clickEdit = useCallback(() => {
    setDisabled(false);
    if (clear) setValue('');
  }, [clear]);

  const clickCancel = useCallback(() => {
    setDisabled(true);
    setValue(defaultValue);
  }, [defaultValue]);

  const clickSave = useCallback(() => {
    save(valueKey ? { [valueKey]: value } : value);
    setDisabled(true);
  }, [save, value, valueKey]);

  return (
    <div className="font-light editable-text-input">
      {label && <div className="mb-4">{label}</div>}
      <div className="border border-purple-200 rounded p-4 flex">
        <input
          value={value}
          onChange={onChange}
          disabled={disabled}
          type={type}
          className="flex-1 p-2 rounded font-light"
        />
        {!fixed && (
          disabled
            ? (
              <button onClick={clickEdit} className={`${className} editable-text-input__edit-btn`} type="button">
                <BiPencil size="30" />
              </button>
            )
            : (
              <>
                <button onClick={clickCancel} className={`${className} editable-text-input__cancel-btn`} type="button"><BiX size="30" /></button>
                <button onClick={clickSave} className={`${className} editable-text-input__save-btn`} type="button"><TiTick size="30" /></button>
              </>
            )
        )}
      </div>
    </div>
  );
}

EditableTextInput.defaultProps = {
  type: 'text',
  clear: false,
  fixed: false,
  save: () => {},
  valueKey: '',
  defaultValue: '',
  label: '',
};

EditableTextInput.propTypes = {
  save: PropTypes.func,
  defaultValue: PropTypes.string,
  label: PropTypes.string,
  type: PropTypes.string,
  clear: PropTypes.bool,
  fixed: PropTypes.bool,
  valueKey: PropTypes.string,
};

export default EditableTextInput;
