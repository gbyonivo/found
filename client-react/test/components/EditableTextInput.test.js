import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import EditableTextInput from '../../src/components/EditableTextInput';

describe('EditableTextInput', () => {
  afterEach(() => {
    jest.clearAllMocks();
  })
  test('should mount properly with default props', () => {
    const onClick = jest.fn(() => {});
    const element = renderer.create(<EditableTextInput save={onClick} />);
    expect(element).toMatchSnapshot();
  });

  test.each([
    ['', 'potato'],
    ['name', { name: 'potato' }],
  ])('clickEdit/onChange/clickSave with valueKey as %p', (valueKey, calledWith) => {
    const onClick = jest.fn(() => {});
    const element = shallow(<EditableTextInput save={onClick} valueKey={valueKey} />);
    element.find('.editable-text-input__edit-btn').simulate('click');
    element.find('input').simulate('change', { target: { value: 'potato' } });
    element.find('.editable-text-input__save-btn').simulate('click');
    expect(element.find('input').prop('value')).toBe('potato');
    expect(onClick).toHaveBeenCalledWith(calledWith);
    expect(element.find('input').prop('disabled')).toBe(true);
  });

  test('clickEdit/onChange/clickCancel', () => {
    const onClick = jest.fn(() => {});
    const element = shallow(<EditableTextInput save={onClick} />);
    element.find('.editable-text-input__edit-btn').simulate('click');
    element.find('input').simulate('change', { target: { value: 'try me' } });
    element.find('.editable-text-input__cancel-btn').simulate('click');
    expect(element.find('input').prop('value')).toBe('');
    expect(element.find('input').prop('disabled')).toBe(true);
  });
});
