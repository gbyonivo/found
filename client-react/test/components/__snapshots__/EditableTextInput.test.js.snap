// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTextInput should mount properly with default props 1`] = `
<div
  className="font-light editable-text-input"
>
  
  <div
    className="border border-purple-200 rounded p-4 flex"
  >
    <input
      className="flex-1 p-2 rounded font-light"
      disabled={true}
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold p-2 rounded inline-flex items-center ml-4 editable-text-input__edit-btn"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        height="30"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 21a1 1 0 0 0 .24 0l4-1a1 1 0 0 0 .47-.26L21 7.41a2 2 0 0 0 0-2.82L19.42 3a2 2 0 0 0-2.83 0L4.3 15.29a1.06 1.06 0 0 0-.27.47l-1 4A1 1 0 0 0 3.76 21 1 1 0 0 0 4 21zM18 4.41 19.59 6 18 7.59 16.42 6zM5.91 16.51 15 7.41 16.59 9l-9.1 9.1-2.11.52z"
        />
      </svg>
    </button>
  </div>
</div>
`;
